apps/gateway/test/champions_test.exs

```diff
@@ -0,0 +1,286 @@
+      assert unit_to_unselect.selected
+
+      # Unselect the unit
+      :ok = SocketTester.unselect_unit(socket_tester, user.id, unit_to_unselect.id)
+      fetch_last_message(socket_tester)
+      assert_receive %WebSocketResponse{response_type: {:unit, %Unit{}}}
+
+      fetch_last_message(socket_tester)
+      %WebSocketResponse{response_type: {:unit, unselected_unit}} = get_last_message()
+
+      assert not unselected_unit.selected
+      # Protobuf doesn't support nil values, returns zero instead
+      assert unselected_unit.slot == 0
+
+      :ok = SocketTester.select_unit(socket_tester, user.id, unselected_unit.id, slot)
+      fetch_last_message(socket_tester)
+      assert_receive %WebSocketResponse{response_type: {:unit, %Unit{}}}
+
+      fetch_last_message(socket_tester)
+      %WebSocketResponse{response_type: {:unit, selected_unit}} = get_last_message()
+
+      assert selected_unit.selected
+      assert selected_unit.slot == slot
+    end
+
+    test "unit progression", %{socket_tester: socket_tester} do
+      muflus = GameBackend.Units.Characters.get_character_by_name("Muflus")
+      {:ok, user} = Users.register("LevelUpUser")
+      Currencies.add_currency_by_name!(user.id, "Gold", 9999)
+
+      {:ok, unit} =
+        GameBackend.Units.insert_unit(%{
+          user_id: user.id,
+          level: 19,
+          tier: 1,
+          rank: 2,
+          selected: false,
+          character_id: muflus.id
+        })
+
+      gold = Currencies.get_amount_of_currency_by_name(user.id, "Gold")
+      level = unit.level
+
+      #### Level up
+      [%CurrencyCost{currency_id: _gold_id, amount: level_up_cost}] = Units.calculate_level_up_cost(unit)
+
+      :ok = SocketTester.level_up_unit(socket_tester, user.id, unit.id)
+      fetch_last_message(socket_tester)
+
+      assert_receive %WebSocketResponse{
+        response_type: {:unit_and_currencies, %UnitAndCurrencies{}}
+      }
+
+      fetch_last_message(socket_tester)
+
+      %WebSocketResponse{
+        response_type: {:unit_and_currencies, %UnitAndCurrencies{unit: unit, user_currency: [user_currency]}}
+      } = get_last_message()
+
+      assert unit.level == level + 1
+      assert user_currency.currency.name == "Gold"
+      assert user_currency.amount == gold - level_up_cost
+
+      # Cannot level up because unit is level 20 with tier 1
+      assert unit.level == 20
+      assert unit.tier == 1
+      :ok = SocketTester.level_up_unit(socket_tester, user.id, unit.id)
+      fetch_last_message(socket_tester)
+      assert_receive %WebSocketResponse{response_type: {:error, %Error{reason: "cant_level_up"}}}
+
+      #### Tier up
+
+      user_gold = Currencies.get_amount_of_currency_by_name(user.id, "Gold")
+      user_gems = Currencies.get_amount_of_currency_by_name(user.id, "Gems")
+
+      [%CurrencyCost{currency_id: _gold_id, amount: gold_cost}, %CurrencyCost{currency_id: _gems_id, amount: gems_cost}] =
+        Units.calculate_tier_up_cost(unit)
```

Comment:
When we implement the real calculate_tier_up_cost there may be a random error here since maybe the currency costs are shuffle and the gem cost appears first

Reply:
Yes, is that a bad thing? When we implement actual tier up costs, we should then implement tests that fit the current implementation at that point, right?